### PR TITLE
Save Card CheckBox

### PR DESCRIPTION
**Pull request by scannillo**
- Added a "Save card" CheckBox to `CardForm` page.
- Has ability to be set as visible or hidden, and as pre-checked, or un-checked.

#### WAITING: For design sign off.

### CODE REVIEW

File: CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java [hunk 1]
```diff
@@ -244,6 +248,25 @@ public CardForm maskCvv(boolean mask) {
         return this;
     }
 
+    /**
+     * @param visible {@code true} to show the Save Card CheckBox, {@code false} to hide the Save Card CheckBox. Defaults to {@code false}.
```
Comment (quinnjn) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java]:
```
@param visible Determines if the save card CheckBox should be shown. Defaults to hidden / {@code false}
```

File: CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java [hunk 2]
```diff
@@ -727,6 +753,14 @@ public String getMobileNumber() {
         return mMobileNumber.getMobileNumber();
     }
 
+    /**
+     * @return the value for whether or not the Save Card CheckBox is checked
+     */
+    public boolean getSaveCardCheckBoxValue() {
```
Comment (quinnjn) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java]: I think this name would be better: 
```
isSaveCardCheckBoxChecked()
```

File: CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java [hunk 3]
```diff
@@ -727,6 +753,14 @@ public String getMobileNumber() {
         return mMobileNumber.getMobileNumber();
     }
 
+    /**
+     * @return the value for whether or not the Save Card CheckBox is checked
```
Comment (quinnjn) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/CardForm.java]: @return whether or not the save card CheckBox is checked

File: CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import android.os.Parcelable;
+import android.util.AttributeSet;
+
+import androidx.appcompat.widget.AppCompatCheckBox;
+
+/**
+ * Sets an initial CheckBox checked state that is overwritten when restoring this view.
+ */
+public class InitialValueCheckBox extends AppCompatCheckBox {
+    private static final String EXTRA_SUPER_STATE = "com.braintreepayments.cardform.view.DefaultValueCheckBox.EXTRA_SUPER_STATE";
```
Comment (quinnjn) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java]: The value should be: `com.braintreepayments.cardform.view.InitialValueCheckBox.EXTRA_SUPER_STATE`

File: CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    private static final String EXTRA_CHECKED_VALUE = "com.braintreepayments.cardform.view.DefaultValueCheckBox.EXTRA_CHECKED_VALUE";
```
Comment (quinnjn) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java]: The value should be: `com.braintreepayments.cardform.view.InitialValueCheckBox.EXTRA_CHECKED_VALUE`

File: CardForm/src/main/res/layout/bt_card_form_fields.xml [hunk 6]
```diff
@@ -170,4 +170,20 @@
...
+    <LinearLayout
+        style="@style/bt_field_container"
+        android:paddingRight="0dp"
+        android:paddingLeft="@dimen/bt_mobile_number_explanation_padding"
+        android:paddingStart="@dimen/bt_mobile_number_explanation_padding"
+        android:paddingEnd="0dp">
+
+        <com.braintreepayments.cardform.view.InitialValueCheckBox
+            android:id="@+id/bt_card_form_save_card_checkbox"
+            android:layout_marginLeft="-7dp"
```
Comment (quinnjn) [inline CardForm/src/main/res/layout/bt_card_form_fields.xml]: This should be moved to a dimen value: `@dimen/bt_save_card_checkbox_left_margin`

File: CardForm/src/main/res/layout/bt_card_form_fields.xml [hunk 7]
```diff
@@ -170,4 +170,20 @@
...
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/bt_save_card_checkbox_name"/>
+
+    </LinearLayout>
+
```
Comment (quinnjn) [inline CardForm/src/main/res/layout/bt_card_form_fields.xml]: This `\n` can be deleted

File: CardForm/src/main/res/layout/bt_card_form_fields.xml [hunk 8]
```diff
@@ -170,4 +170,20 @@
...
+        android:paddingStart="@dimen/bt_mobile_number_explanation_padding"
+        android:paddingEnd="0dp">
+
+        <com.braintreepayments.cardform.view.InitialValueCheckBox
+            android:id="@+id/bt_card_form_save_card_checkbox"
+            android:layout_marginLeft="-7dp"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/bt_save_card_checkbox_name"/>
+
```
Comment (quinnjn) [inline CardForm/src/main/res/layout/bt_card_form_fields.xml]: This `\n` can be deleted

File: CardForm/src/test/java/com/braintreepayments/cardform/view/TestInitialValueCheckBox.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        assertTrue(initialValueCheckBox.isChecked());
+
+        Parcelable parcelable = initialValueCheckBox.onSaveInstanceState();
+        initialValueCheckBox.onRestoreInstanceState(parcelable);
+
+        initialValueCheckBox.setInitiallyChecked(false);
+
+        assertTrue(initialValueCheckBox.isChecked());
+    }
+
```
Comment (quinnjn) [inline CardForm/src/test/java/com/braintreepayments/cardform/view/TestInitialValueCheckBox.java]: This `\n` can be deleted

File: Sample/src/main/java/com/braintreepayments/sample/BaseCardFormActivity.java [hunk 10]
```diff
@@ -81,4 +83,5 @@ public boolean onOptionsItemSelected(MenuItem item) {
 
         return false;
     }
+
```
Comment (quinnjn) [inline Sample/src/main/java/com/braintreepayments/sample/BaseCardFormActivity.java]: This `\n` can be deleted

File: CardForm/build.gradle [hunk 11]
```diff
@@ -3,7 +3,7 @@ apply plugin: 'maven'
 apply plugin: 'signing'
 apply plugin: 'io.codearte.nexus-staging'
 
-version '4.1.1-SNAPSHOT'
+version '4.1.1-DEVELOPMENT'
```
Comment (quinnjn) [inline CardForm/build.gradle]: This change should be reverted

File: CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import android.os.Bundle;
+import android.os.Parcelable;
+import android.util.AttributeSet;
+
+import androidx.appcompat.widget.AppCompatCheckBox;
+
+/**
+ * Sets an initial CheckBox checked state that is overwritten when restoring this view.
+ */
+public class InitialValueCheckBox extends AppCompatCheckBox {
```
Comment (scannillo) [inline CardForm/src/main/java/com/braintreepayments/cardform/view/InitialValueCheckBox.java]: Didn't we decide on toggle? https://braintree.slack.com/archives/GB6RZSQV9/p1555624166178500

**Pull request merged**